src/lib/spinel/radio_spinel_impl.hpp

```diff
@@ -1838,5 +2055,138 @@ uint32_t RadioSpinel<InterfaceType, ProcessContextType>::GetBusSpeed(void) const
     return mSpinelInterface.GetBusSpeed();
 }
 
+template <typename InterfaceType, typename ProcessContextType>
+void RadioSpinel<InterfaceType, ProcessContextType>::HandleRcpUnexpectedReset(void)
+{
+#if OPENTHREAD_SPINEL_CONFIG_RCP_RESTORATION_ENABLE
+    mRcpFailed = true;
+#else
+    DieNow(OT_EXIT_RADIO_SPINEL_RESET);
+#endif
+}
+
+template <typename InterfaceType, typename ProcessContextType>
+void RadioSpinel<InterfaceType, ProcessContextType>::HandleRcpTimeout(void)
+{
+#if OPENTHREAD_SPINEL_CONFIG_RCP_RESTORATION_ENABLE
+    mRcpFailed = true;
+#else
+    DieNow(OT_EXIT_RADIO_SPINEL_NO_RESPONSE);
+#endif
+}
+
+template <typename InterfaceType, typename ProcessContextType>
+void RadioSpinel<InterfaceType, ProcessContextType>::RecoverFromRcpFailure(void)
+{
+#if OPENTHREAD_SPINEL_CONFIG_RCP_RESTORATION_ENABLE
+    if (!mRcpFailed)
+    {
+        ExitNow();
+    }
+    mRcpFailed = false;
+
+    if (mIsRcpRecovering)
+    {
+        otLogCritPlat("RCP failure detected during RCP recovering, exiting.");
+        DieNow(OT_EXIT_FAILURE);
+    }
+    otLogWarnPlat("RCP failure detected, trying to recover.");
+    mIsRcpRecovering = true;
+
+    mRcpRecoveringState = mState;
+
+    mState = kStateDisabled;
+    mRxFrameBuffer.Clear();
+    mSpinelInterface.OnRcpReset();
+    mCmdTidsInUse = 0;
+    mCmdNextTid   = 1;
+    mTxRadioTid   = 0;
+    mWaitingTid   = 0;
+    mWaitingKey   = SPINEL_PROP_LAST_STATUS;
+    mIsReady      = false;
```

Comment:
When do we set `mIsReady = true` after reset restoration?

Reply:
In `HandleValueIs`: <https://github.com/openthread/openthread/blob/7bdc83f8b3afe782dc01a231a51b79579a9aab11/src/lib/spinel/radio_spinel_impl.hpp#L793>

Reply:
This statement is skipped if `HandleRcpUnexpectedReset()` is called.  Do we have another chance to set `mIsReady` when restoring?

Reply:
`HandleRcpUnexpectedReset` only marks the failure state. Then when the actual recovery happens `mIsReady` will be set to false first, then set to true when RCP's first message arrives.